cshared/main.go

```diff
@@ -0,0 +1,117 @@
+// This package is intended to be compiled into a C shared library for
+// use by non-Golang clients to perform certificate and signing operations.
+//
+// The shared library exports language-specific wrappers around the Golang
+// client APIs.
+//
+// Example compilation command:
+// go build -buildmode=c-shared -o signer.dylib main.go
+package main
+
+/*
+#include <stdlib.h>
+*/
+import "C"
+
+import (
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/rsa"
+	"encoding/pem"
+	"log"
+	"unsafe"
+
+	"github.com/googleapis/enterprise-certificate-proxy/client"
+)
+
+func getCertPem(configFilePath string) []byte {
+	key, err := client.Cred(configFilePath)
+	if err != nil {
+		log.Printf("Could not create client using config %s: %v", configFilePath, err)
+		return nil
+	}
+	defer key.Close()
+
+	certChain := key.CertificateChain()
+	certChainPem := []byte{}
+	for i := 0; i < len(certChain); i++ {
+		certPem := pem.EncodeToMemory(&pem.Block{Type: "CERTIFICATE", Bytes: certChain[i]})
+		certChainPem = append(certChainPem, certPem...)
+	}
+	return certChainPem
+}
+
+//export GetCertPemForPython
+//
+// GetCertPemForPython reads the contents of the certificate specified by configFilePath,
+// storing the result inside a certHolder byte array of size certHolderLen.
+//
+// We must call it twice to get the cert. First time use nil for certHolder to get
+// the cert length. Second time we pre-create an array in Python of the cert length and
+// call this function again to load the cert into the array.
+func GetCertPemForPython(configFilePath *C.char, certHolder *byte, certHolderLen int) int {
+	pemBytes := getCertPem(C.GoString(configFilePath))
+	if certHolder != nil {
+		cert := unsafe.Slice(certHolder, certHolderLen)
+		copy(cert, pemBytes)
+	}
+	return len(pemBytes)
+}
+
+//export SignForPython
+//
+// SignForPython signs a message digest of length digestLen using a certificate private key
+// specified by configFilePath, storing the result inside a sigHolder byte array of size sigHolderLen.
+func SignForPython(configFilePath *C.char, digest *byte, digestLen int, sigHolder *byte, sigHolderLen int) int {
```

Comment:
should we check that sigHolderLen is 256? so that we can avoid weird memory issues?

Reply:
when it is used in google-auth lib, I am passing in sigHolder buffer of length 2000, which is more than enough. This function will fill the signature into the buffer, and return the actual length of the signature. I made a suggestion to add a size check.

Reply:
Applied your suggestion. Thanks!